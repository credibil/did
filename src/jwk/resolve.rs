//! Resolution of a DID into a public key using the `did:jwk` method.

use base64ct::{Base64UrlUnpadded, Encoding};

use crate::{Error, Method, MethodType, Resource, Url, VerificationMethodBuilder, VmKeyId};

/// Convert a `did:jwk` URL into a [`VerificationMethod`] object.
/// 
/// # Errors
/// If the URL is not a valid `did:jwk` URL, an error is returned. 
pub fn resolve(url: &Url) -> crate::Result<Resource> {
    if url.method != Method::Jwk {
        return Err(Error::InvalidDid(format!("DID is not a valid did:jwk: {url}")));
    }
    // For JWK, the method-specific identifier is the encoded key.
    let decoded = Base64UrlUnpadded::decode_vec(&url.id)
        .map_err(|e| Error::InvalidDid(format!("issue decoding key: {e}")))?;
    let jwk = serde_json::from_slice(&decoded)
        .map_err(|e| Error::InvalidDid(format!("issue deserializing key: {e}")))?;

    let vm = VerificationMethodBuilder::new(&jwk)
        .key_id(&url.resource_id(), VmKeyId::Index(String::new(), 0))?
        .method_type(&MethodType::JsonWebKey2020)?
        .build();
    
    Ok(Resource::VerificationMethod(vm))
}