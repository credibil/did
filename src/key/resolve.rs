//! Resolution of a DID into a public key using the `did:jwk` method.

use crate::{Error, Method, MethodType, PublicKeyFormat, Resource, Url, VerificationMethod};

/// Convert a `did:key` URL into a [`VerificationMethod`] object.
/// 
/// # Errors
/// If the URL is not a valid `did:key` URL, an error is returned.
pub fn resolve(url: &Url) -> crate::Result<Resource> {
    if url.method != Method::Key {
        return Err(Error::InvalidDid(format!("DID is not a valid did:key: {url}")));
    }
    // For did:key, the fragment is the key already multibase encoded. There is
    // no need to use a builder.
    let Some(fragment) = &url.fragment else {
        return Err(Error::InvalidDid("DID is not a valid did:key - there is no fragment".into()));
    };
    let vm = VerificationMethod {
        context: None,
        id: url.resource_id(),
        type_: MethodType::Multikey,
        controller: url.did(),
        key: PublicKeyFormat::PublicKeyMultibase {
            public_key_multibase: fragment.to_string(),
        },
    };
    Ok(Resource::VerificationMethod(vm))
}
