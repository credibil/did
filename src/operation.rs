//! Helpers for DID operations that are independent of DID method.

pub mod document;
pub mod resolve;
