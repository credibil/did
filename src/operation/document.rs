//! # DID Document helpers.
//!
//! This module provides helpers to create and manipulate DID Documents
//! independent of the DID method.
//!
//! See [Decentralized Identifiers (DIDs) v1.0](https://www.w3.org/TR/did-1.0/)
//! for more information.
//!

use anyhow::{anyhow, bail};
use base64ct::{Base64UrlUnpadded, Encoding};
use credibil_infosec::PublicKeyJwk;
use curve25519_dalek::edwards::CompressedEdwardsY;
use multibase::Base;
use serde_json::Value;

use crate::{
    KeyPurpose,
    core::{Kind, OneMany},
};

use crate::document::{
    Document, DocumentMetadata, MethodType, PublicKeyFormat, Service, VerificationMethod,
};

/// A builder for creating a DID Document.
#[derive(Clone, Debug, Default)]
pub struct DocumentBuilder<O> {
    /// Operation being performed
    pub operation: O,

    // Document under construction
    doc: Document,
}

// Typestate state guards for a `DocumentBuilder`.

/// The `DocumentBuilder` is being used to create new DID document.
#[derive(Default)]
pub struct Create;
/// The `DocumentBuilder` is being used to update an existing DID document.
#[derive(Default)]
pub struct Update;

impl<O> DocumentBuilder<O> {
    /// Creates a new `DocumentBuilder` with the given DID URL.
    #[must_use]
    pub fn new(did: &str) -> DocumentBuilder<Create> {
        let doc = Document {
            id: did.to_string(),
            ..Document::default()
        };
        DocumentBuilder { operation: Create, doc }
    }

    /// Creates a new `DocumentBuilder` from an existing `Document`.
    #[must_use]
    pub fn from(doc: Document) -> DocumentBuilder<Update> {
        DocumentBuilder {
            operation: Update,
            doc,
        }
    }

    /// Add an also-known-as identifier.
    #[must_use]
    pub fn also_known_as(mut self, aka: &str) -> Self {
        self.doc.also_known_as.get_or_insert(vec![]).push(aka.to_string());
        self
    }

    /// Add a controller.
    ///
    /// Chain to add multiple controllers.
    #[must_use]
    pub fn controller(mut self, controller: &str) -> Self {
        match self.doc.controller {
            Some(c) => match c {
                OneMany::One(cont) => {
                    self.doc.controller = Some(OneMany::Many(vec![cont, controller.to_string()]));
                }
                OneMany::Many(mut cont) => {
                    cont.push(controller.to_string());
                    self.doc.controller = Some(OneMany::Many(cont));
                }
            },
            None => {
                self.doc.controller = Some(OneMany::One(controller.to_string()));
            }
        }
        self
    }

    /// Add a service endpoint.
    ///
    /// Chain to add multiple service endpoints.
    #[must_use]
    pub fn service(mut self, service: &Service) -> Self {
        self.doc.service.get_or_insert(vec![]).push(service.clone());
        self
    }

    /// Add a context.
    ///
    /// Chain to add multiple contexts.
    #[must_use]
    pub fn context(mut self, context: &Kind<Value>) -> Self {
        self.doc.context.push(context.clone());
        self
    }

    /// Add a verification method.
    ///
    /// Pass the ID of the verification method to associate with the
    /// relationship or a complete `VerificationMethod` instance if a standalone
    /// method is required.
    ///
    /// Chain to add multiple relationships.
    ///
    /// # Errors
    ///
    /// If this method is used for key agreement, an error will occur if a
    /// standalone verification method is not used.
    pub fn verification_method(
        mut self, vm: &Kind<VerificationMethod>, purpose: &KeyPurpose,
    ) -> anyhow::Result<Self> {
        match purpose {
            KeyPurpose::Authentication => {
                self.doc.authentication.get_or_insert(vec![]).push(vm.clone());
            }
            KeyPurpose::AssertionMethod => {
                self.doc.assertion_method.get_or_insert(vec![]).push(vm.clone());
            }
            KeyPurpose::KeyAgreement => match vm {
                Kind::Object(vm) => {
                    self.doc.key_agreement.get_or_insert(vec![]).push(Kind::Object(vm.clone()));
                }
                Kind::String(_) => {
                    return Err(anyhow!(
                        "key agreement must be handled by the encryption method creation algorithm"
                    ));
                }
            },
            KeyPurpose::CapabilityInvocation => {
                self.doc.capability_invocation.get_or_insert(vec![]).push(vm.clone());
            }
            KeyPurpose::CapabilityDelegation => {
                self.doc.capability_delegation.get_or_insert(vec![]).push(vm.clone());
            }
            KeyPurpose::VerificationMethod => match vm {
                Kind::Object(vm) => {
                    self.doc.verification_method.get_or_insert(vec![]).push(vm.clone());
                }
                Kind::String(_) => {
                    return Err(anyhow!(
                        "verification method must be a standalone verification method"
                    ));
                }
            },
        }
        Ok(self)
    }

    /// Retrieve the current `DID` from the builder.
    ///
    /// Note that although the `DID` (document identifier) is called for in the
    /// constructor of this builder, some DID methods may use temporary values
    /// and replace the DID in the final document. Users of this function should
    /// be aware of the DID method context in which it is used to determine the
    /// reliability of the value.
    #[must_use]
    pub fn did(&self) -> &str {
        &self.doc.id
    }
}

impl<Create> DocumentBuilder<Create> {
    /// Set default metadata with created timestamp and build the DID Document.
    #[must_use]
    pub fn build(mut self) -> Document {
        let md = DocumentMetadata {
            created: chrono::Utc::now(),
            ..Default::default()
        };
        self.doc.did_document_metadata = Some(md);
        self.doc
    }
}

impl<Update> DocumentBuilder<Update> {
    /// Construct a new DID `Document` with updated timestamp.
    #[must_use]
    pub fn update(mut self) -> Document {
        let mut md = self.doc.did_document_metadata.clone().unwrap_or_default();
        md.updated = Some(chrono::Utc::now());
        self.doc.did_document_metadata = Some(md);
        self.doc
    }
}

/// A builder for creating a verification method.
#[derive(Default)]
pub struct VerificationMethodBuilder {
    vm_key: PublicKeyJwk,
    did: String,
    kid: String,
    method: MethodType,
    format: PublicKeyFormat,
}

impl VerificationMethodBuilder {
    /// Creates a new `VerificationMethodBuilder` with the given public key.
    #[must_use]
    pub fn new(verifying_key: &PublicKeyJwk) -> Self {
        Self {
            vm_key: verifying_key.clone(),
            ..Default::default()
        }
    }

    /// Specify how to construct the key ID.
    ///
    /// # Errors
    ///
    /// Will fail if the ID type requires a multibase value but construction of
    /// that value fails.
    pub fn key_id(mut self, did: &str, id_type: VmKeyId) -> anyhow::Result<Self> {
        self.did = did.to_string();
        match id_type {
            VmKeyId::Authorization(auth_key) => {
                let mb = auth_key.to_multibase()?;
                self.kid = format!("{did}#{mb}");
            }
            VmKeyId::Verification => {
                let mb = self.vm_key.to_multibase()?;
                self.kid = format!("{did}#{mb}");
            }
            VmKeyId::Index(index) => {
                self.kid = format!("{did}#key-{index}");
            }
        }
        Ok(self)
    }

    /// Specify the verification method type.
    ///
    /// To generate an `X25519` public encryption key from an `Ed25519` key, use
    /// [`derive_key_agreement`] instead of this function.
    ///
    /// # Errors
    ///
    /// Will fail if required format is multibase but the public key cannot be
    /// decoded into bytes.
    /// TODO: Is there a nicer way to handle this? This type mapping is
    /// awkward.
    pub fn method_type(mut self, mtype: &MethodType) -> anyhow::Result<Self> {
        self.method = mtype.clone();
        self.format = match mtype {
            MethodType::Multikey
            | MethodType::Ed25519VerificationKey2020
            | MethodType::X25519KeyAgreementKey2020 => {
                // multibase encode the public key
                PublicKeyFormat::PublicKeyMultibase {
                    public_key_multibase: self.vm_key.to_multibase()?,
                }
            }
            MethodType::JsonWebKey2020 | MethodType::EcdsaSecp256k1VerificationKey2019 => {
                PublicKeyFormat::PublicKeyJwk {
                    public_key_jwk: self.vm_key.clone(),
                }
            }
        };
        Ok(self)
    }

    /// Special case for specifying the public key format: derive a `X25519` key
    /// agreement verification method from an `Ed25519` key.
    ///
    /// # Errors
    ///
    /// Will fail if decoding or computing the Edwards Curve point fails.
    pub fn derive_key_agreement(mut self) -> anyhow::Result<Self> {
        // derive an X25519 public encryption key from the Ed25519 key
        let key_bytes = Base64UrlUnpadded::decode_vec(&self.vm_key.x)?;
        let edwards_y = CompressedEdwardsY::from_slice(&key_bytes)?;
        let Some(edwards_point) = edwards_y.decompress() else {
            bail!("Edwards Y cannot be decompressed to a point");
        };
        let x25519_bytes = edwards_point.to_montgomery().to_bytes();
        self.method = MethodType::X25519KeyAgreementKey2020;
        self.format = PublicKeyFormat::PublicKeyMultibase {
            public_key_multibase: multibase::encode(Base::Base58Btc, x25519_bytes),
        };
        Ok(self)
    }

    /// Build the verification method.
    #[must_use]
    pub fn build(self) -> VerificationMethod {
        VerificationMethod {
            id: self.kid,
            controller: self.did,
            type_: self.method,
            key: self.format,
            ..VerificationMethod::default()
        }
    }
}

/// Instruction to the `VerificationMethodBuilder` on how to construct the key
/// ID.
pub enum VmKeyId {
    /// Use the provided authorization key and construct a multibase value from
    /// that to append to the document identifier (DID URL).
    Authorization(PublicKeyJwk),

    /// Use the verification method key from the `DidOperator` to construct a
    /// multibase value to append to the document identifier (DID URL).
    Verification,

    /// Increment an index to append to the document identifier (DID URL).
    ///
    /// `key-0`, `key-1`, etc.
    Index(u32),
}
