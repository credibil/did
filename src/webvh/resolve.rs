//! # DID Web with Verifiable History Resolver
//!
//! Resolution of a DID for the `did:webvh` method.
//!
//! See: <https://identity.foundation/didwebvh/next/>

use std::{sync::LazyLock, vec};

use anyhow::anyhow;
use chrono::{DateTime, Utc};
use multibase::Base;
use regex::Regex;
use serde_json::json;

use super::{
    DidLogEntry, WitnessEntry,
    verify::{verify_proofs, verify_witness},
};
use crate::operation::resolve::{ContentType, Metadata, Options, Parameters, Resolved};
use crate::{DidResolver, Document, Error, webvh::SCID_PLACEHOLDER};

static DID_REGEX: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new("^did:webvh:(?<identifier>[a-zA-Z0-9.\\-:\\%]+)$").expect("should compile")
});

/// Resolve a `did:webvh` DID URL to a DID document.
///
/// The first step of the resolution is to retrieve and parse the DID list
/// document. See further functions in this implementation to help with
/// resolution steps.
///
/// # Errors
///
/// Will fail if the DID URL is invalid or the DID list document cannot be
/// found.
pub async fn resolve(
    did: &str, options: Option<Options>, resolver: impl DidResolver,
) -> crate::Result<Resolved> {
    // Generate the URL to fetch the DID list document.
    let url = http_url(did, None)?;

    // The content type for the did.jsonl file SHOULD be text/jsonl.
    if let Some(opts) = options {
        if let Some(content_type) = opts.accept {
            match content_type {
                ContentType::JsonL => {}
                ContentType::DidLdJson => {
                    return Err(Error::RepresentationNotSupported(
                        "Content type must be text/json".to_string(),
                    ));
                }
            }
        }
    }

    // Perform an HTTP GET request to the URL for the DID log document. The
    // client can use helper methods to unpack the `JSONL` file and extract the
    // DID document.
    // TODO: The resolver needs to honour the content type.
    let document = resolver.resolve(&url).await.map_err(Error::Other)?;

    Ok(Resolved {
        context: "https://w3id.org/did-resolution/v1".into(),
        metadata: Metadata {
            content_type: ContentType::DidLdJson,
            additional: Some(json!({
                "pattern": "^did:webvh:(?<identifier>[a-zA-Z0-9.\\-:\\%]+)$",
                "did": {
                    "didString": did,
                    "methodSpecificId": did[8..],
                    "method": "webvh"
                }
            })),
            ..Metadata::default()
        },
        document: Some(document),
        ..Resolved::default()
    })
}

/// Convert a `did:webvh` URL to an HTTP URL pointing to the location of the
/// DID list document (default) or another location where the root path is
/// a conversion from the DID to an HTTP URL.
///
/// # Errors
///
/// Will fail if the DID URL is invalid.
///
/// <https://identity.foundation/didwebvh/#the-did-to-https-transformation>
///
fn http_url(did: &str, file_path: Option<&str>) -> crate::Result<String> {
    let Some(caps) = DID_REGEX.captures(did) else {
        return Err(Error::InvalidDid("DID is not a valid did:webvh".to_string()));
    };
    // 1. Remove the literal `did:webvh:` prefix from the DID URL.
    let scid_and_fqdn = &caps["identifier"];

    // 2. Remove the `SCID` by removing the text up to and including the
    // first `:` character.
    let Some(fqdn) = scid_and_fqdn.split_once(':').map(|x| x.1) else {
        return Err(Error::InvalidDid("DID is not a valid did:webvh - no SCID".to_string()));
    };

    // 3. Replace `:` with `/` in the domain part of the identifier to obtain
    // the fully qualified domain name and optional path.
    let mut domain = fqdn.replace(':', "/");

    // 4. If there is no optional path, append `/.well-known` to the URL.
    if !fqdn.contains(':') {
        domain.push_str("/.well-known");
    }

    // 5. If the domain contains a port, percent-decode the colon.
    let domain = domain.replace("%3A", ":");

    // 6. Prepend `https://` to the domain to generate the URL.
    let url = format!("https://{domain}");

    // 7. Append `/did.jsonl` (default) or the specified file sub-path to
    // the URL to complete it.
    let fp = file_path.unwrap_or("/did.jsonl");
    let url = format!("{url}{fp}");

    Ok(url)
}

/// Primary verification of the contents of the `did.jsonl` file.
///
/// To use this function, read the contents of the `did.jsonl` file into a
/// vector of `DidLogEntry` structs and pass to this function.
///
/// # Errors
///
/// Will fail if the log entries are invalid.
///
pub async fn resolve_log(
    log: &[DidLogEntry], witnesses: Option<&[WitnessEntry]>, parameters: Option<Parameters>,
    resolver: &impl DidResolver,
) -> crate::Result<Document> {
    let mut prev_index = 0;
    let mut prev_version = SCID_PLACEHOLDER.to_string();
    let mut prev_time = DateTime::<Utc>::MIN_UTC;
    let mut doc = Document::default();
    let mut prev_next_key_hashes: Option<Vec<String>> = None;
    for i in 0..log.len() {
        // 1. Update current parameters with parameters from the entry being
        // processed.

        // 2. Verify controller proofs.
        if let Err(error) = verify_proofs(&log[i], resolver).await {
            return Err(Error::Other(error));
        }

        // 3.1. Verify the version number is incremented by one for each entry.
        // 3.2. Verify the version number and entry hash is separated by `-`.
        let version_parts = log[i].version_id.split('-').collect::<Vec<&str>>();
        if version_parts.len() != 2 {
            return Err(Error::Other(anyhow!("log entry version id has an unexpected format")));
        }
        let index = version_parts[0].parse::<u64>().map_err(|e| Error::Other(e.into()))?;
        if index != prev_index + 1 {
            return Err(Error::Other(anyhow!("log entries are not sequential")));
        }

        // 3.3. Verify the entry hash.
        log[i].verify_hash(&prev_version).map_err(Error::Other)?;

        // 4. The version time must be in the past and monotonically increasing.
        if log[i].version_time > Utc::now() {
            return Err(Error::Other(anyhow!("log entry time is in the future")));
        }
        if log[i].version_time <= prev_time {
            return Err(Error::Other(anyhow!("log entry times are not monotonically increasing")));
        }

        // 5. If the entry is the first one, verify the SCID.
        if i == 0 {
            let mut initial_log_entry = log[i].clone();
            initial_log_entry.version_id = SCID_PLACEHOLDER.to_string();
            initial_log_entry.proof = vec![];
            let hash = initial_log_entry.hash().map_err(Error::Other)?;
            if hash != log[i].parameters.scid {
                return Err(Error::Other(anyhow!(
                    "first log entry SCID does not match calculated hash"
                )));
            }
        }

        // 6. Record the state as the document to return (if everything else is
        // successful).
        doc = log[i].state.clone();

        // 7. If key pre-rotation is enabled, check the update keys match the
        // previous entry's next-key hashes.
        if let Some(next_key_hashes) = &prev_next_key_hashes {
            for key in &log[i].parameters.update_keys {
                let key_hash = multibase::encode(Base::Base58Btc, key.as_bytes());
                if !next_key_hashes.contains(&key_hash) {
                    return Err(Error::Other(anyhow!(
                        "update key not found in pre-rotation hashes"
                    )));
                }
            }
        }

        // 8. Increment.
        prev_index = index;
        prev_version.clone_from(&log[i].version_id);
        prev_time.clone_from(&log[i].version_time);
        prev_next_key_hashes.clone_from(&log[i].parameters.next_key_hashes);

        // 9. Check witness proofs.
        if log[i].parameters.witness.is_some() {
            if let Some(witness_entries) = witnesses {
                if let Err(error) = verify_witness(&log[i], witness_entries, resolver).await {
                    return Err(Error::Other(error));
                }
            } else {
                return Err(Error::Other(anyhow!(
                    "witness proofs must be provided for a log entry with witnesses"
                )));
            }
        }

        // Check for explicit version ID or version time request. (Otherwise
        // the latest version is returned.)
        if let Some(params) = &parameters {
            if let Some(version_id) = &params.version_id {
                if *version_id == log[i].version_id {
                    break;
                }
            }
            if let Some(version_time) = &params.version_time {
                let version_time =
                    version_time.parse::<DateTime<Utc>>().map_err(|e| Error::Other(e.into()))?;
                if version_time >= log[i].version_time {
                    if i < log.len() - 1 {
                        if version_time <= log[i + 1].version_time {
                            break;
                        }
                    } else {
                        break;
                    }
                }
            }
        }
    }
    Ok(doc)
}

#[cfg(test)]
mod test {
    use anyhow::anyhow;
    use insta::assert_json_snapshot as assert_snapshot;

    use crate::{Document, operation::resolve::dereference};

    use super::*;

    #[derive(Clone)]
    struct MockResolver;
    impl DidResolver for MockResolver {
        async fn resolve(&self, _url: &str) -> anyhow::Result<Document> {
            serde_json::from_slice(include_bytes!("./did-v3.json"))
                .map_err(|e| anyhow!("issue deserializing document: {e}"))
        }
    }

    #[tokio::test]
    async fn deref_webvh() {
        const DID_URL: &str = "did:webvh:QmaJp6pmb6RUk4oaDyWQcjeqYbvxsc3kvmHWPpz7B5JwDU:credibil.io#z6MkijyunEqPi7hzgJirb4tQLjztCPbJeeZvXEySuzbY6MLv";

        let dereferenced =
            dereference(DID_URL, None, MockResolver).await.expect("should dereference");
        assert_snapshot!("deref_webvh", dereferenced);
    }

    #[test]
    fn should_construct_default_url() {
        let did =
            "did:webvh:QmaJp6pmb6RUk4oaDyWQcjeqYbvxsc3kvmHWPpz7B5JwDU:domain.with-hyphens.computer";
        let url = http_url(did, None).unwrap();
        assert_eq!(url, "https://domain.with-hyphens.computer/.well-known/did.jsonl");
    }

    #[test]
    fn should_construct_path_url() {
        let did = "did:webvh:QmaJp6pmb6RUk4oaDyWQcjeqYbvxsc3kvmHWPpz7B5JwDU:domain.with-hyphens.computer:dids:issuer";
        let url = http_url(did, None).unwrap();
        assert_eq!(url, "https://domain.with-hyphens.computer/dids/issuer/did.jsonl");
    }

    #[test]
    fn should_construct_port_url() {
        let did = "did:webvh:QmaJp6pmb6RUk4oaDyWQcjeqYbvxsc3kvmHWPpz7B5JwDU:domain.with-hyphens.computer%3A8080";
        let url = http_url(did, None).unwrap();
        assert_eq!(url, "https://domain.with-hyphens.computer:8080/.well-known/did.jsonl");
    }
}
