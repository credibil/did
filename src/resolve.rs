//! # DID Resolver
//!
//! This crate provides a DID Resolver trait and a set of default
//! implementations for resolving DIDs.
//!
//! See [DID resolution](https://www.w3.org/TR/did-core/#did-resolution) fpr more.

use std::str::FromStr;

use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::document::{Document, Service, VerificationMethod};
use crate::error::Error;
use crate::{DidResolver, Method, Url, key, web, webvh};

/// Dereference a DID URL into a resource.
///
/// If you have destructured DID URL already, you can bypass this function and
/// call `deref_url` directly. See [`deref_url`] for more information.
///
/// # Errors
/// Will return an error if the DID URL cannot be parsed or the provided
/// resolver fails to resolve the source DID document.
///
/// Note that only URLs implying DID methods supported by this crate will
/// survive parsing.
pub async fn dereference(
    did_url: &str, resolver: &impl DidResolver,
) -> crate::Result<Resource> {
    let url = crate::url::Url::from_str(did_url)?;
    deref_url(&url, resolver).await
}

/// Dereference a structured DID URL into a resource.
///
/// Construct a structured DID URL using the `Url` struct for the `url`
/// parameter or call `dereference_url` with a string to do the same.
///
/// For DID documents that are hosted (on a web server, in a ledger, etc.), you
/// will need to implement the `DidResolver` trait for your resolver. to resolve
/// the DID document from the `Url`. This crate will then do its best to find
/// the requested resource in the document.
///
/// For self-describing DIDs a full document is never resolved. The
/// `DidResolver` can be a no-op because it's `resolve` method is not called.
/// For example, the `did:key` method can return a public key from the DID URL
/// fragment.
///
/// # Errors
/// Will return an error if the provided resolver fails to resolve the source
/// DID document.
///
/// Will also return an error if the resource is not found in the document. This
/// includes cases that don't make sense, like asking a `did:key` for a service
/// endpoint.
pub async fn deref_url(url: &Url, resolver: &impl DidResolver) -> crate::Result<Resource> {
    match url.method {
        Method::Key => key::resolve(url),
        Method::Web => {
            let doc = web::resolve(url, resolver).await?;
            document_resource(url, &doc)
        }
        Method::WebVh => {
            let doc = webvh::resolve(url, resolver).await?;
            document_resource(url, &doc)
        }
    }
}

/// Get a resource from a DID document.
///
/// Uses the `Url` to infer the type of resource to return.
///
/// # Errors
/// Will return an error if the resource is not found in the document.
pub fn document_resource(url: &Url, doc: &Document) -> crate::Result<Resource> {
    if let Some(query) = &url.query {
        if let Some(service_id) = &query.service {
            if let Some(service) = doc.get_service(service_id) {
                return Ok(Resource::Service(service.clone()));
            }
            return Err(Error::NotFound(format!("service {service_id} not found in document")));
        }
    }
    if url.fragment.is_none() {
        return Ok(Resource::Document(doc.clone()));
    }
    if let Some(vm) = doc.get_verification_method(&url.to_string()) {
        return Ok(Resource::VerificationMethod(vm.clone()));
    }
    Err(Error::NotFound(format!("verification method {url} not found in document")))
}

/// Resource represents the DID document resource returned as a result of DID
/// dereferencing. The resource is a DID document or a subset of a DID document.
#[allow(clippy::large_enum_variant)]
#[derive(Clone, Debug, Deserialize, Serialize)]
pub enum Resource {
    ///  DID `Document` resource.
    Document(Document),

    /// `VerificationMethod` resource.
    VerificationMethod(VerificationMethod),

    /// `Service` resource.
    Service(Service),
}

impl Default for Resource {
    fn default() -> Self {
        Self::VerificationMethod(VerificationMethod::default())
    }
}

/// DID document metadata.
#[derive(Clone, Debug, Default, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Metadata {
    /// The Media Type of the returned resource.
    pub content_type: ContentType,

    /// The error code from the dereferencing process, if applicable.
    /// Values of this field SHOULD be registered in the DID Specification
    /// Registries. Common values are `invalid_did_url` and `not_found`.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,

    /// A human-readable explanation of the error.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error_message: Option<String>,

    /// Additional information about the resolution or dereferencing process.
    #[serde(flatten)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub additional: Option<Value>,
}

/// The Media Type of the returned resource.
#[derive(Clone, Debug, Default, Deserialize, Serialize)]
pub enum ContentType {
    /// JSON-LD representation of a DID document.
    #[default]
    #[serde(rename = "application/did+ld+json")]
    DidLdJson,
    
    /// The JSON-LD Media Type.
    #[serde(rename = "application/ld+json")]
    LdJson,

    /// JSON list document.
    #[serde(rename = "text/jsonl")]
    JsonL,
}
