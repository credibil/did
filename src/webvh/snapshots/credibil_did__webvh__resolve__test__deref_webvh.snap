---
source: src/webvh/resolve.rs
assertion_line: 174
expression: dereferenced
---
{
  "metadata": {
    "contentType": "application/did+ld+json"
  },
  "contentStream": {
    "VerificationMethod": {
      "id": "did:webvh:QmaJp6pmb6RUk4oaDyWQcjeqYbvxsc3kvmHWPpz7B5JwDU:credibil.io#z6MkijyunEqPi7hzgJirb4tQLjztCPbJeeZvXEySuzbY6MLv",
      "controller": "did:webvh:QmaJp6pmb6RUk4oaDyWQcjeqYbvxsc3kvmHWPpz7B5JwDU:credibil.io",
      "type": "Multikey",
      "publicKeyMultibase": "z6MkijyunEqPi7hzgJirb4tQLjztCPbJeeZvXEySuzbY6MLv"
    }
  },
  "contentMetadata": {}
}
